x/wasm/internal/keeper/recurse_test.go

```diff
@@ -0,0 +1,259 @@
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type Recurse struct {
+	Depth    uint32         `json:"depth"`
+	Work     uint32         `json:"work"`
+	Contract sdk.AccAddress `json:"contract"`
+}
+
+type recurseWrapper struct {
+	Recurse Recurse `json:"recurse"`
+}
+
+func buildQuery(t *testing.T, msg Recurse) []byte {
+	wrapper := recurseWrapper{Recurse: msg}
+	bz, err := json.Marshal(wrapper)
+	require.NoError(t, err)
+	return bz
+}
+
+type recurseResponse struct {
+	Hashed []byte `json:"hashed"`
+}
+
+func TestGasCostOnQuery(t *testing.T) {
+	const (
+		GasNoWork uint64 = InstanceCost + 2_756
+		// Note: about 100 SDK gas (10k wasmer gas) for each round of sha256
+		GasWork50 uint64 = InstanceCost + 8_464 // this is a little shy of 50k gas - to keep an eye on the limit
+
+		GasReturnUnhashed uint64 = 647
+		GasReturnHashed   uint64 = 597
+	)
+
+	cases := map[string]struct {
+		gasLimit    uint64
+		msg         Recurse
+		expectedGas uint64
+	}{
+		"no recursion, no work": {
+			gasLimit:    400_000,
+			msg:         Recurse{},
+			expectedGas: GasNoWork,
+		},
+		"no recursion, some work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Work: 50, // 50 rounds of sha256 inside the contract
+			},
+			expectedGas: GasWork50,
+		},
+		"recursion 1, no work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Depth: 1,
+			},
+			expectedGas: 2*GasNoWork + GasReturnUnhashed,
+		},
+		"recursion 1, some work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Depth: 1,
+				Work:  50,
+			},
+			expectedGas: 2*GasWork50 + GasReturnHashed,
+		},
+		"recursion 4, some work": {
+			gasLimit: 400_000,
+			msg: Recurse{
+				Depth: 4,
+				Work:  50,
+			},
+			// this is (currently) 244_708 gas
+			expectedGas: 5*GasWork50 + 4*GasReturnHashed,
+		},
+	}
+
+	// we do one basic setup before all test cases (which are read-only and don't change state)
+	tempDir, err := ioutil.TempDir("", "wasm")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	ctx, keepers := CreateTestInput(t, false, tempDir, SupportedFeatures, nil, nil)
+	accKeeper, keeper := keepers.AccountKeeper, keepers.WasmKeeper
+	deposit := sdk.NewCoins(sdk.NewInt64Coin("denom", 100000))
+	creator := createFakeFundedAccount(ctx, accKeeper, deposit.Add(deposit...))
+
+	// store the code
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+	codeID, err := keeper.Create(ctx, creator, wasmCode, "", "", nil)
+	require.NoError(t, err)
+
+	// instantiate the contract
+	_, _, bob := keyPubAddr()
+	_, _, fred := keyPubAddr()
+	initMsg := InitMsg{
+		Verifier:    fred,
+		Beneficiary: bob,
+	}
+	initMsgBz, err := json.Marshal(initMsg)
+	require.NoError(t, err)
+	contractAddr, err := keeper.Instantiate(ctx, codeID, creator, nil, initMsgBz, "recursive contract", deposit)
+	require.NoError(t, err)
+
+	for name, tc := range cases {
+		tc := tc
```

Comment:
nit: This is not needed

Reply:
Oh yes it is.. at least with the golangci linter on the cosmos-sdk.

I guess I can keep it out of here, but am a bit traumatized by that linter (and the long feedback cycle)